src/hotspot/cpu/aarch64/spin_wait_aarch64.hpp

```diff
@@ -25,24 +25,57 @@
 #ifndef CPU_AARCH64_SPIN_WAIT_AARCH64_HPP
 #define CPU_AARCH64_SPIN_WAIT_AARCH64_HPP
 
-class SpinWait {
+// SpinWait provides a description for implementations of spin wait/pause.
+// The description includes:
+// - what an instruction should be used by an implementation.
+// - how many of the instructions.
+// - a runner which can execute the requested number of instructions.
+//
+// Creation of SpinWait is controlled by VM_Version.
+class SpinWait final {
 public:
   enum Inst {
     NONE = -1,
     NOP,
     ISB,
     YIELD
   };
+  using InstRunner = void (*)(int count);
 
 private:
   Inst _inst;
   int _count;
+  InstRunner _inst_runner;
 
-public:
-  SpinWait(Inst inst = NONE, int count = 0) : _inst(inst), _count(count) {}
+  static void run_nop(int count) {
+    while (count-- > 0) {
+      __asm volatile("nop");
+    }
+  }
+
+  static void run_isb(int count) {
+    while (count-- > 0) {
+      __asm volatile("isb");
+    }
+  }
+
+  static void run_yield(int count) {
+    while (count-- > 0) {
+      __asm volatile("yield");
+    }
+  }
+
+  static void run_none(int) {}
 
+  SpinWait(Inst inst = NONE, int count = 0, InstRunner inst_runner = run_none) :
+      _inst(inst), _count(count), _inst_runner(inst_runner)  {}
```

Comment:
Wouldn't it make more sense to have _inst_runner initialized in the constructor based on the value of Inst inst?
You aren't differentiating between the two in get_spin_wait_desc anyway.